components/movementsensor/adxl345/adxl345.go

```diff
@@ -232,6 +331,123 @@ func (adxl *adxl345) writeByte(ctx context.Context, register, value byte) error
+	register := getByteFromInterrupts(adxl.interruptsEnabled)
+	return adxl.writeByte(ctx, IntEnable, register)
+}
+
+func (adxl *adxl345) setRelevantRegisters(ctx context.Context) error {
+	if len(adxl.interruptsEnabled) == 0 {
+		return nil
+	}
+	relevantRegisters := adxl.getRelevantRegisterValues()
+	for key := range defaultRegisterValues {
+		val, ok := relevantRegisters[key]
+		if ok {
+			err := adxl.writeByte(ctx, key, val)
+			if err != nil {
+				return errors.Errorf("Unable to write byte at register %x", key)
+			}
+		} else {
+			err := adxl.writeByte(ctx, key, defaultRegisterValues[key])
+			if err != nil {
+				return errors.Errorf("Unable to write byte at register %x", key)
+			}
+		}
+	}
+	return nil
+}
+
+func (adxl *adxl345) getAxes(x, y, z bool) byte {
+	var tapAxes byte
+	if x {
+		tapAxes += xBit
+	}
+	if y {
+		tapAxes += yBit
+	}
+	if z {
+		tapAxes += zBit
+	}
+	return tapAxes
+}
+
+func (adxl *adxl345) getRelevantRegisterValues() map[byte]byte {
+	relevantRegisterValues := map[byte]byte{}
+
+	tapAxesSpecified := adxl.tapX || adxl.tapY || adxl.tapZ
+	if tapAxesSpecified {
+		relevantRegisterValues[TapAxes] = adxl.getAxes(adxl.tapX, adxl.tapY, adxl.tapZ)
+	}
+
+	var actInactAxes byte
+	activityAxesSpecified := adxl.actX || adxl.actY || adxl.actZ
+	if activityAxesSpecified {
+		actInactAxes = adxl.getAxes(adxl.actX, adxl.actY, adxl.actZ)
+	}
+	inactivityAxesSpecified := adxl.inactX || adxl.inactY || adxl.inactZ
+	if inactivityAxesSpecified {
+		actInactAxes += (adxl.getAxes(adxl.inactX, adxl.inactY, adxl.inactZ) << 4)
+	}
+	if activityAxesSpecified || inactivityAxesSpecified {
+		relevantRegisterValues[ActInactCtl] = actInactAxes
+	}
+
+	if adxl.threshTap != 0 {
+		relevantRegisterValues[ThreshTap] = adxl.threshTap
+	}
+	if adxl.dur != 0 {
+		relevantRegisterValues[Dur] = adxl.dur
+	}
+	if adxl.latent != 0 {
+		relevantRegisterValues[Latent] = adxl.latent
+	}
+	if adxl.window != 0 {
+		relevantRegisterValues[Window] = adxl.window
+	}
+	if adxl.threshFf != 0 {
+		relevantRegisterValues[ThreshFf] = adxl.threshFf
+	}
+	if adxl.timeFf != 0 {
+		relevantRegisterValues[TimeFf] = adxl.timeFf
+	}
+	if adxl.threshAct != 0 {
+		relevantRegisterValues[ThreshAct] = adxl.threshAct
+	}
+	if adxl.threshInact != 0 {
+		relevantRegisterValues[ThreshInact] = adxl.threshInact
+	}
+	if adxl.timeInact != 0 {
+		relevantRegisterValues[TimeInact] = adxl.timeInact
+	}
+	if adxl.actInactCtl != 0 {
+		relevantRegisterValues[ActInactCtl] = adxl.timeFf
+	}
+	return relevantRegisterValues
+}
+
+func (adxl *adxl345) readInterrupts(ctx context.Context) {
+	interuptEnabledRegister := getByteFromInterrupts(adxl.interruptsEnabled)
+	intSourceRegister, err := adxl.readByte(ctx, IntSource)
+	if err != nil {
+		adxl.logger.Error(err)
+	}
+
+	for key, value := range interruptBitPosition {
+		if intSourceRegister&value&interuptEnabledRegister != 0 {
+			adxl.interruptsFound = append(adxl.interruptsFound, key)
```

Comment:
Part of me says it's great to parse this stuff out because as a human it's easier to read this data structure, and part of me says this isn't so great because it's slow and the user should choose whether they want it parsed out like this or not (perhaps the thing that consumes this data goes through extra work just to reconstruct the original byte again, and we should just provide the raw data). What do other folks think?

Reply:
Fair point - I could go either way. It would also be easy enough to send both